src/components/frontend-engine/frontend-engine.tsx

```diff
@@ -1,73 +1,72 @@
 import { yupResolver } from "@hookform/resolvers/yup";
-import { useCallback, useEffect, useState } from "react";
-import { Controller, useForm } from "react-hook-form";
+import { forwardRef, useEffect, useImperativeHandle, useState } from "react";
+import { Controller, FormProvider, useForm } from "react-hook-form";
 import { useValidationSchema } from "src/utils/hooks";
-import * as FrontendEngineFields from "..";
-import { StyledForm } from "./frontend-engine.styles";
-import { FieldType, IFrontendEngineProps, TFrontendEngineFieldSchema } from "./types";
+import * as FrontendEngineFields from "../fields";
+import {
+	FieldType,
+	IFrontendEngineProps,
+	IFrontendEngineRef,
+	IGenericFieldProps,
+	TFrontendEngineFieldSchema,
+	TFrontendEngineValues,
+} from "./types";
 import { ValidationProvider } from "./validation-schema";
 
-const FrontendEngineInner = (props: IFrontendEngineProps) => {
-	// ================================================
+const FrontendEngineInner = forwardRef<IFrontendEngineRef, IFrontendEngineProps>((props, ref) => {
+	// =============================================================================
 	// CONST, STATE, REFS
-	// ================================================
+	// =============================================================================
 	const {
-		id,
-		className = "",
-		data,
-		defaultValues,
-		validators,
-		conditions,
-		validationMode,
-		revalidationMode,
+		data: {
+			className: dataClassName,
+			defaultValues,
+			fields: fieldData,
+			id,
+			revalidationMode = "onChange",
+			validationMode = "onSubmit",
+		},
+		className,
 		onSubmit,
-		onValidate,
 	} = props;
 
 	const [fields, setFields] = useState<JSX.Element[]>([]);
 	const { validationSchema } = useValidationSchema();
 
-	const {
-		control,
-		watch,
-		handleSubmit: reactFormHookSubmit,
-		formState: { errors },
-	} = useForm({
+	const formMethods = useForm({
 		mode: validationMode,
-		reValidateMode: revalidationMode || "onChange",
+		reValidateMode: revalidationMode,
 		defaultValues: defaultValues,
 		resolver: yupResolver(validationSchema),
 	});
+	const { control, watch, handleSubmit: reactFormHookSubmit, formState } = formMethods;
 
-	// ================================================
+	// =============================================================================
 	// EFFECTS
-	// ================================================
```

Comment:
note: moved `buildFieldsFromSchema` to be called directly in `useEffect` to avoid `useCallback` (unnecessary memoisation)